src/API/Endpoints/Reports/Income.php

```diff
@@ -0,0 +1,134 @@
+	}
+
+	public function get_report( $request ) {
+
+		// Check if a cached version exists
+		$cached_report = $this->get_cached_report( $request );
+		if ( $cached_report !== false ) {
+			// Bail and return the cached version
+			return new \WP_REST_Response(
+				[
+					'data' => $cached_report,
+				]
+			);
+		}
+
+		$start = date_create( $request['start'] );
+		$end   = date_create( $request['end'] );
+		$diff  = date_diff( $start, $end );
+
+		$data = [];
+
+		switch ( true ) {
+			case ( $diff->days > 900 ):
+				$data = $this->get_data( $start, $end, 'P1Y', 'Y' );
+				break;
+			case ( $diff->days > 700 ):
+				$data = $this->get_data( $start, $end, 'P6M', 'F Y' );
+				break;
+			case ( $diff->days > 400 ):
+				$data = $this->get_data( $start, $end, 'P3M', 'F Y' );
+				break;
+			case ( $diff->days > 120 ):
+				$data = $this->get_data( $start, $end, 'P1M', 'M Y' );
+				break;
+			case ( $diff->days > 30 ):
+				$data = $this->get_data( $start, $end, 'P7D', 'M jS' );
+				break;
+			case ( $diff->days > 10 ):
+				$data = $this->get_data( $start, $end, 'P3D', 'M jS' );
+				break;
+			case ( $diff->days > 4 ):
+				$data = $this->get_data( $start, $end, 'P1D', 'l' );
+				break;
+			case ( $diff->days > 1 ):
+				$data = $this->get_data( $start, $end, 'P1D', 'D ga' );
+				break;
+			case ( $diff->days >= 0 ):
+				$data = $this->get_data( $start, $end, 'PT1H', 'D ga' );
+				break;
+		}
+
+		// Cache the report data
+		$result = $this->cache_report( $request, $data );
+
+		return new \WP_REST_Response(
+			[
+				'data' => $data,
+			]
+		);
+	}
+
+	public function get_data( $start, $end, $interval, $format ) {
+
+		$stats = new \Give_Payment_Stats();
+
+		$startStr = $start->format( 'Y-m-d H:i:s' );
+		$endStr   = $end->format( 'Y-m-d H:i:s' );
+
+		// Determine the start date of the previous period (used to calculate trend)
+		$prev    = date_sub( date_create( $startStr ), date_diff( $start, $end ) );
+		$prevStr = $prev->format( 'Y-m-d H:i:s' );
+
+		$labels = [];
+		$income = [];
+
+		$dateInterval = new \DateInterval( $interval );
+		while ( $start < $end ) {
+
+			$periodStart = $start->format( 'Y-m-d H:i:s' );
+
+			// Add interval to get period end
+			$periodEnd = clone $start;
+			date_add( $periodEnd, $dateInterval );
+
+			$label     = $periodEnd->format( $format );
+			$periodEnd = $periodEnd->format( 'Y-m-d H:i:s' );
+
+			$incomeForPeriod = $stats->get_earnings( 0, $periodStart, $periodEnd );
+
+			$income[] = $incomeForPeriod;
+			$labels[] = $label;
+
+			date_add( $start, $dateInterval );
+		}
+
+		$totalForPeriod = array_sum( $income );
+
+		// Calculate the income trend by comparing total earnings in the
+		// previous period to earnings in the current period
+		$prevTotal    = $stats->get_earnings( 0, $prevStr, $startStr );
+		$currentTotal = $stats->get_earnings( 0, $startStr, $endStr );
+		$trend        = $prevTotal > 0 ? round( ( ( $currentTotal - $prevTotal ) / $prevTotal ) * 100 ) : 'NaN';
```

Comment:
Can you explain the result of "Nan"? Is this a scenario you expect? Is the downstream code expecting it?

Reply:
Because you cannot calculate the trend of a metric that starts at zero or negative (ex: if you started with zero donors, and the next month you had 100 donors), but I do expect certain periods could start at zero, it seems appropriate to handle this situation. If a non-numeric value is provided to the MiniChart, it will not display trend indicator/percentage. I chose to with "NaN" since it indicates that a number could not be calculated, but this could just as easily return false or null (implying there is no trend). Open to ideas on return values, and how this fits into our discussions about error handling.

Reply:
Got it! Just wanted to make sure it's not an error scenario, but rather a realistic one that may occur legitimately. I would reach for null in this case as the return, as we're saying "there is no trend". NaN is really a JS value which a string "NaN" does not communicate. Make sense?